### PR TITLE
Dynamic tree transport

### LINKED ISSUE #56: Add option to show the whole tree in visualization
## The problem
When debugging a DSD with the visualization software it is sometimes useful to know what else *could* happen if some decision evaluated differently.

Also when the DSD does eventually evaluate differently, the window jumps around because the graph changes shape.

## Desired solution
It would be nice if there was an option to show the whole DSD in visualization. The path not taken should still be marked in light gray.

### LINKED ISSUE #69: DSD GUI uses a whole core
The dsd GUI uses a whole core when we view the body behavior with its 125 Hz.

### LINKED ISSUE #102: Visualization: Automatically detect packages with DSD
# Summary
The DSD visualization should automatically detect the packages which use a DSD and show them in the GUI instead of relying on a hard-coded configuration file. 

## Is your feature request related to a problem?
The configuration file must be updated whenever other packages are restructured and cannot be shipped independently of these packages. 

## Describe the solution you'd like
There could be a marker in the package.xml file or a config file with the files could be added at a certain location. I think there are other pieces of software with ROS that work in such a way. Maybe the MoveIt plugin framework?

### PR DESCRIPTION
**Pull request by Flova**
# Summary
- No hard-coded DSD paths in the VIZ anymore
- No second DSD running in the viz node
- Tree json serialization
- Provide the tree on a latched topic
- Hot reload if DSD tree changes
- The viz subscribes to that topic and discovers DSDs that are currently active automatically using it
- Non hardcoded ROS params are now displayed properly
- Option to show the full tree (kinda slow for large trees, need to optimize that)
- Fix highlight button
- Disable fit to view if the user scrolls once
- Better caching (it actually does something now)
- Typing annotations 
- Backward compatible viz for ROS Bags that include the latched tree topic in the future if the local files change 
- Some color changes

Closes #102 
Closes #56 
Closes #69

### CODE REVIEW

**Comment by Flova** [comment]
We might need to check that the UDP bridge and ros bag recordings work with latched topics and also record or transmit the correct topics.

File: dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_visualization_plugin.py [hunk 1]
```diff
@@ -106,24 +100,74 @@ def toggle_freeze():
...
+        # Add to InteractiveGraphicsView mouse wheel event that disables auto-fit-in-view
+        def wheel_event(event):
+            self._widget.auto_fit_graph_check_box.setChecked(False)
+            self._widget.graphics_view.__class__.wheelEvent(self._widget.graphics_view, event)
+
+        # Call it as well as the original wheelEvent
+        self._widget.graphics_view.wheelEvent = wheel_event
+
+        # Add widget to the user interface
+
```
Comment (jaagut) [inline dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_visualization_plugin.py]:
```suggestion
```

File: dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_visualization_plugin.py [hunk 2]
```diff
@@ -106,24 +100,74 @@ def toggle_freeze():
...
+        self._widget.auto_fit_graph_check_box.toggled.connect(reset_dot_cache)
+        self._widget.fit_in_view_push_button.pressed.connect(reset_dot_cache)
+        self._widget.freeze_push_button.toggled.connect(reset_dot_cache)
+
+        # Add to InteractiveGraphicsView mouse wheel event that disables auto-fit-in-view
+        def wheel_event(event):
+            self._widget.auto_fit_graph_check_box.setChecked(False)
+            self._widget.graphics_view.__class__.wheelEvent(self._widget.graphics_view, event)
+
+        # Call it as well as the original wheelEvent
```
Comment (jaagut) [inline dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_visualization_plugin.py]:
```suggestion
        # Overwrite the original wheelEvent handler
```

File: dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py [hunk 3]
```diff
@@ -1,112 +1,112 @@
...
+        # Subscribe to the DSDs tree (latched)
+        self.tree_sub = self._node.create_subscription(
+            String,
+            f"{debug_topic}/dsd_tree",
+            self._tree_callback,
+            qos_profile=QoSProfile(depth=10, durability=DurabilityPolicy.TRANSIENT_LOCAL),
+        )
 
-            self._parse_remote_data(remaining_data["next"], self.tree.root_element)
+        self._node.get_logger().info(f"Subscribed to {debug_topic}/dsd_tree")
```
Comment (jaagut) [inline dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py]: please create a string variable with the full topic name and use it for the subscription + logging. That way, we can not forget to change only one occurence in the future. Also applies to the second subscription.

File: dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py [hunk 4]
```diff
@@ -135,184 +135,223 @@ def _empty_item_model():
...
+            output = []
+            for param_name, param_value in params.items():
+                output.append(f"{param_name}: {str(param_value)}")
+            return " (" + ", ".join(output) + ")"
+
+        # Sanity check
+        if stack_element is not None:
+            assert stack_element["type"] == tree_element["type"], "The stack and the tree do not match"
+            if stack_element["type"] != "sequence":
+                assert stack_element["name"] == tree_element["name"], "The stack and the tree do not match"
```
Comment (jaagut) [inline dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py]: Please log for which element it does not match

File: dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py [hunk 5]
```diff
@@ -135,184 +135,223 @@ def _empty_item_model():
...
+        :param subtree_root: The root node of the subtree which should be added to the graph
+        :param stack_root: The root node / bottom of the stack which should be added to the graph (this corresponds to subtree_root as they are different representations of the same component). The stack is optional as certain subtrees might not be in contact with the stack.
+        :param full_tree: If true the whole tree is rendered, otherwise only the current stack and its direct neighbors are rendered
+        :return: The modified dotgraph and the uid of the root node
+        """
+        # Sanity check
+        if stack_root is not None:
+            assert stack_root["type"] == subtree_root["type"], "The stack and the tree do not match"
+            if stack_root["type"] != "sequence":
+                assert stack_root["name"] == subtree_root["name"], "The stack and the tree do not match"
```
Comment (jaagut) [inline dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py]:
```suggestion
            assert stack_root["type"] == subtree_root["type"], "The stack root type and the tree root type do not match"
            if stack_root["type"] != "sequence":
                assert stack_root["name"] == subtree_root["name"], "The stack root name and the tree root name do not match"
```

File: dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py [hunk 6]
```diff
@@ -135,184 +135,223 @@ def _empty_item_model():
...
+        # Generate dot node for the root and mark it as active if it is on the stack
+        dot_node = DsdFollower._dot_node_from_tree_element(subtree_root, stack_root)
+        # Append this element to graph
+        dot.add_node(dot_node)
+
+        # Append children to graph if this element is on the stack or if we want to render the whole tree
+        if "children" in subtree_root and (stack_root is not None or full_tree):
+            # Go through all children
+            for activating_result, child in subtree_root["children"].items():
+                # Get the root of the sub stack if this branch the one which is currently on the stack
```
Comment (jaagut) [inline dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py]: Grammar?

File: dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py [hunk 7]
```diff
@@ -135,184 +135,223 @@ def _empty_item_model():
...
+        return dot, dot_node.get_name()
 
-        return dot, node.get_name()
-
-    def _append_element_to_item(self, parent_item, debug_data):
+    def _append_debug_data_to_item(
+        self, parent_item: QStandardItem, debug_data: Union[dict, list, int, float, str, bool]
+    ):
         """
         Append an elements debug_data to a QStandardItem.
```
Comment (jaagut) [inline dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py]: Grammar?

File: dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py [hunk 8]
```diff
@@ -135,184 +135,223 @@ def _empty_item_model():
...
 
-    def load_behavior(self, path):
+    def destroy(self):
         """
-        Load a .dsd file into the behaviour to execute it. This should be called after the actions
-        and decisions have been loaded.
-        :param path: The path to the .dsd file describing the behaviour
-        :return:
+        Cleanup the subscriptions, to we don't receive any more data that we don't need (performance)
+        Also this allows the garbage collector to delete this object and prevent memory leaks
```
Comment (jaagut) [inline dynamic_stack_decider_visualization/dynamic_stack_decider_visualization/dsd_follower.py]:
```suggestion
        Cleanup the subscriptions, so we don't receive any more data that we don't need.
        This improves the performance.
        Also this allows the garbage collector to delete this object and prevent memory leaks.
```

**Review by jaagut — APPROVED** [review]

**Pull request merged**